File: watcher-web/src/api/knowledge.ts
import request from '@/utils/system/request'

export interface KnowledgeBase {
  id: string
  name: string
  description: string
  status: string
  document_count: number
  chunk_count: number
  created_at: string
  updated_at: string
}

export interface Document {
  id: string
  kb_id: string
  file_name: string
  file_path: string
  file_size: number
  status: string
  chunk_count: number
  created_at: string
}

export interface ChatMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface ChatRequest {
  kb_id: string
  question: string
  history?: ChatMessage[]
}

export interface ChatResponse {
  answer: string
  sources?: Array<{
    content: string
    file_name: string
    score: number
  }>
}

// 对话历史相关类型
export interface ChatSession {
  id: string
  user_id: string
  kb_id: string
  title: string
  message_count: number
  created_at: string
  updated_at: string
}

export interface ChatHistoryItem {
  id: number
  session_id: string
  role: string
  content: string
  sources: any
  created_at: string
}

export interface ChatHistoryListResponse {
  sessions: ChatSession[]
  total: number
  page: number
  page_size: number
}

export interface SessionDetailResponse {
  session: ChatSession
  messages: ChatHistoryItem[]
}

export interface SearchRequest {
  query: string
  top_k?: number
}

export interface SearchResult {
  documents: Array<{
    content: string
    file_name: string
    chunk_index: number
    score: number
  }>
}

// 知识库 CRUD
export async function listKnowledgeBases(): Promise<KnowledgeBase[]> {
  const res = await request.get<KnowledgeBase[]>('/api/knowledge/kbs')
  return res.data
}

export async function createKnowledgeBase(data: { name: string; description?: string }): Promise<KnowledgeBase> {
  const res = await request.post<KnowledgeBase>('/api/knowledge/kbs', data)
  return res.data
}

export async function getKnowledgeBase(id: string): Promise<KnowledgeBase> {
  const res = await request.get<KnowledgeBase>(`/api/knowledge/kbs/${id}`)
  return res.data
}

export async function deleteKnowledgeBase(id: string): Promise<void> {
  await request.delete(`/api/knowledge/kbs/${id}`)
}

export async function buildKnowledgeBase(id: string): Promise<void> {
  await request.post(`/api/knowledge/kbs/${id}/build`)
}

export async function getKnowledgeBaseStats(id: string): Promise<{ document_count: number; chunk_count: number; status: string }> {
  const res = await request.get(`/api/knowledge/kbs/${id}/stats`)
  return res.data
}

// 文档管理
export async function uploadDocument(kbId: string, file: File): Promise<Document> {
  const formData = new FormData()
  formData.append('file', file)
  const res = await request.post<Document>(`/api/knowledge/kbs/${kbId}/documents`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
  return res.data
}

export async function listDocuments(kbId: string): Promise<Document[]> {
  const res = await request.get<Document[]>(`/api/knowledge/kbs/${kbId}/documents`)
  return res.data
}

export async function deleteDocument(kbId: string, docId: string): Promise<void> {
  await request.delete(`/api/knowledge/kbs/${kbId}/documents/${docId}`)
}

// RAG 问答
export async function chatWithKB(data: ChatRequest): Promise<ChatResponse> {
  const res = await request.post<ChatResponse>('/api/knowledge/chat', data)
  return res.data
}

export async function searchKB(kbId: string, query: string, top_k: number = 3): Promise<SearchResult> {
  const res = await request.post<SearchResult>(`/api/knowledge/kbs/${kbId}/search`, {
    query,
    top_k
  })
  return res.data
}

// 对话历史 API
export async function listChatHistory(userId: string, page: number = 1, pageSize: number = 20): Promise<ChatHistoryListResponse> {
  const res = await request.get<ChatHistoryListResponse>('/api/knowledge/chat/history', {
    params: { user_id: userId, page, page_size: pageSize }
  })
  return res.data
}

export async function getSessionDetail(sessionId: string, userId: string): Promise<SessionDetailResponse> {
  const res = await request.get<SessionDetailResponse>(`/api/knowledge/chat/sessions/${sessionId}`, {
    params: { user_id: userId }
  })
  return res.data
}

export async function deleteSession(sessionId: string, userId: string): Promise<void> {
  await request.delete(`/api/knowledge/chat/sessions/${sessionId}`, {
    params: { user_id: userId }
  })
}

export async function chatWithHistory(data: {
  user_id: string
  kb_id: string
  question: string
  session_id?: string
}): Promise<{ answer: string; session_id: string }> {
  const res = await request.post('/api/knowledge/chat/with-history', data)
  return res.data
}

// 按知识库查询会话列表（用于历史会话弹窗）
export async function getChatSessions(kbId: string): Promise<ChatHistoryListResponse> {
  const res = await request.get<ChatHistoryListResponse>('/api/knowledge/chat/history', {
    params: { kb_id: kbId, page: 1, page_size: 100 }
  })
  return res.data
}

// 删除会话
export async function deleteChatSession(sessionId: string): Promise<void> {
  await request.delete(`/api/knowledge/chat/sessions/${sessionId}`)
}

File: watcher-ai/src/watcher_ai/api/knowledge.py
from fastapi import APIRouter, HTTPException, UploadFile, File
from pydantic import BaseModel
from typing import List, Optional, Any
from watcher_ai.models.schemas import KnowledgeBaseCreate, KnowledgeBaseResponse
from watcher_ai.services.kb_service import KBService
from watcher_ai.services.document_store import DocumentStore
from watcher_ai.services.chroma_service import ChromaService
from watcher_ai.services.build_service import BuildService
from watcher_ai.services.llm_service import LLMService
from watcher_ai.services.chat_service import ChatService
from watcher_ai.services.chat_repository import ChatRepository
import os

router = APIRouter(prefix="/api/knowledge", tags=["knowledge"])

ALLOWED_EXTENSIONS = {".pdf", ".md", ".txt"}

# Request/Response 模型
class ChatRequest(BaseModel):
    kb_id: str
    question: str
    history: Optional[List[dict]] = None
    user_id: Optional[str] = None
    session_id: Optional[str] = None

class SearchRequest(BaseModel):
    query: str
    top_k: int = 3

def ok_response(data: Any):
    """包装成功响应，匹配前端格式"""
    return {"state": 0, "data": data}

@router.post("/kbs", status_code=201)
def create_kb(req: KnowledgeBaseCreate):
    kb = KBService.create(req.name, req.description)
    return ok_response(kb)

@router.get("/kbs")
def list_kbs():
    return ok_response(KBService.list_all())

@router.get("/kbs/{kb_id}")
def get_kb(kb_id: str):
    kb = KBService.get_by_id(kb_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return ok_response(kb)

@router.delete("/kbs/{kb_id}")
def delete_kb(kb_id: str):
    if not KBService.delete(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return ok_response({"message": "deleted"})

# 文档管理
@router.post("/kbs/{kb_id}/documents", status_code=201)
async def upload_document(kb_id: str, file: UploadFile = File(...)):
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    ext = os.path.splitext(file.filename)[1].lower()
    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(status_code=400, detail=f"Unsupported file format: {ext}")
    content = await file.read()
    doc = DocumentStore.save_file(kb_id, file.filename, content)
    return ok_response(doc)

@router.get("/kbs/{kb_id}/documents")
def list_documents(kb_id: str):
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return ok_response(DocumentStore.list_by_kb(kb_id))

@router.delete("/kbs/{kb_id}/documents/{doc_id}")
def delete_document(kb_id: str, doc_id: str):
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    if not DocumentStore.delete(doc_id):
        raise HTTPException(status_code=404, detail="Document not found")
    return ok_response({"message": "deleted"})

# 构建与统计
@router.post("/kbs/{kb_id}/build")
def build_kb(kb_id: str):
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    try:
        result = BuildService.build_knowledge_base(kb_id)
        return ok_response(result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        KBService.update_status(kb_id, "error")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/kbs/{kb_id}/stats")
def get_kb_stats(kb_id: str):
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    kb = KBService.get_by_id(kb_id)
    chunk_count = ChromaService.get_count(kb_id)
    return ok_response({
        "document_count": kb["document_count"],
        "chunk_count": chunk_count,
        "status": kb["status"]
    })

# RAG 问答
@router.post("/chat")
def chat(req: ChatRequest):
    """RAG 问答"""
    kb = KBService.get_by_id(req.kb_id)
    if not kb:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    
    chunk_count = ChromaService.get_count(req.kb_id)
    if chunk_count == 0:
        raise HTTPException(status_code=400, detail="Knowledge base is empty, please build first")
    
    results = ChromaService.search(req.kb_id, req.question, top_k=3)
    
    if not results.get("documents") or not results["documents"][0]:
        # 即使没有检索结果也保存对话
        if req.user_id:
            session_id, message_id = ChatService.create_or_append_session(
                user_id=req.user_id,
                kb_id=req.kb_id,
                question=req.question,
                session_id=req.session_id
            )
            return ok_response({
                "answer": "没有找到相关文档",
                "sources": [],
                "session_id": session_id,
                "message_id": message_id
            })
        return ok_response({"answer": "没有找到相关文档", "sources": []})
    
    context_parts = []
    sources = []
    for doc, meta, dist in zip(
        results["documents"][0],
        results["metadatas"][0],
        results["distances"][0]
    ):
        context_parts.append(f"[{meta.get('file_name', 'unknown')}]: {doc}")
        sources.append({
            "content": doc[:200],
            "file_name": meta.get("file_name", "unknown"),
            "score": 1 - dist
        })
    
    context = "\n\n".join(context_parts)
    answer = LLMService.chat(req.question, context)
    
    # 保存对话历史
    session_id = req.session_id
    message_id = None
    if req.user_id:
        session_id, message_id = ChatService.create_or_append_session(
            user_id=req.user_id,
            kb_id=req.kb_id,
            question=req.question,
            session_id=req.session_id
        )
        # 保存助手回复
        ChatRepository.save_message(
            session_id=session_id,
            role="assistant",
            content=answer,
            sources=sources
        )
        # 更新消息数
        session = ChatRepository.get_session_by_id(session_id)
        ChatRepository.update_message_count(session_id, session['message_count'] + 1)
    
    return ok_response({
        "answer": answer,
        "sources": sources,
        "session_id": session_id,
        "message_id": message_id
    })

@router.post("/kbs/{kb_id}/search")
def search_kb(kb_id: str, req: SearchRequest):
    """检索知识库"""
    if not KBService.get_by_id(kb_id):
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    
    results = ChromaService.search(kb_id, req.query, top_k=req.top_k)
    
    documents = []
    for doc, meta, dist in zip(
        results.get("documents", [[]])[0] if results.get("documents") else [],
        results.get("metadatas", [[]])[0] if results.get("metadatas") else [],
        results.get("distances", [[]])[0] if results.get("distances") else []
    ):
        documents.append({
            "content": doc,
            "file_name": meta.get("file_name", "unknown"),
            "chunk_index": meta.get("chunk_index", 0),
            "score": 1 - dist
        })
    
    return ok_response({"documents": documents})
